### PR TITLE
Remove parameter nullchecking feature

### LINKED ISSUE #61390: Fully remove parameter null-checking implementation
There are a lot of conflicts when trying to automatically revert parameter null-checking in the compiler, and it was done somewhat concurrently with some parser error tolerance improvements, etc. Some combination of automatic and hand-reverting is probably the best way to do it.

We also need to decide what to do with the editor feature to adopt parameter null checking as described in https://github.com/dotnet/runtime/issues/68326#issuecomment-1110157803.

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #61390 

This PR was prepared manually (i.e. without using `git revert`). All changes should be reviewed manually--thankfully many of the changes are sweeping deletes of tests or are fairly boilerplate in nature. Changes in LanguageParser are not so mechanical and should be reviewed more carefully.

This PR does the following:
1. removes related editor features and settings.
2. removes the compiler implementation.
3. changes the parser to add the '!!' as skipped tokens along with a "feature unavailable" diagnostic, instead of adding the '!!' as a "proper" child token of the parameter syntax.

@cston, @chsienki, @CyrusNajmabadi for review

### CODE REVIEW

File: src/VisualStudio/CSharp/Impl/Options/AutomationObject/AutomationObject.Style.cs [hunk 1]
```diff
@@ -244,12 +244,6 @@ public string Style_PreferIsNullCheckOverReferenceEqualityMethod
             set { SetXmlOption(CodeStyleOptions2.PreferIsNullCheckOverReferenceEqualityMethod, value); }
         }
 
-        public string Style_PreferParameterNullChecking
-        {
-            get { return GetXmlOption(CSharpCodeStyleOptions.PreferParameterNullChecking); }
-            set { SetXmlOption(CSharpCodeStyleOptions.PreferParameterNullChecking, value); }
-        }
```
Comment (Youssef1313) [inline src/VisualStudio/CSharp/Impl/Options/AutomationObject/AutomationObject.Style.cs:251]: Does this change hurt backward compatibility?

https://github.com/dotnet/roslyn/blob/3fcc9af35de964c195ff764b32e360d53e4cb5ac/src/VisualStudio/CSharp/Impl/Options/AutomationObject/AutomationObject.ObsoleteAndUnused.cs#L12-L18

If not we could delete `AutomationObject.ObsoleteAndUnused.cs`. I think a good way of testing that is:

1. Export VS settings from the current version which has the option.
2. Make sure the produced file contains this option
3. Try to load the file from a VS build from this PR that doesn't have an option.
4. Will we try access the getter that doesn't exist?

File: src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs [hunk 2]
```diff
@@ -167,8 +167,6 @@ internal static class IDEDiagnosticIds
 
         public const string UseTupleSwapDiagnosticId = "IDE0180";
 
-        public const string UseParameterNullCheckingId = "IDE0190";
-
```
Comment (Youssef1313) [inline src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:171]: Is it safe to reuse IDE0190? If not, consider adding a comment.

File: src/Analyzers/Core/Analyzers/Analyzers.projitems [hunk 3]
```diff
@@ -10,7 +10,9 @@
   </PropertyGroup>
   <ItemGroup>
     <EmbeddedResource Include="$(MSBuildThisFileDirectory)AnalyzersResources.resx" GenerateSource="true" Link="AnalyzersResources.resx" />
-    <None Include="$(MSBuildThisFileDirectory)AnalyzersResources.resx" />
+    <None Include="$(MSBuildThisFileDirectory)AnalyzersResources.resx">
+      <SubType>Designer</SubType>
+    </None>
```
Comment (Youssef1313) [inline src/Analyzers/Core/Analyzers/Analyzers.projitems:15]: Is this change needed/intentional?

**Review by Youssef1313 — COMMENTED** [review]
Consider removing any public APIs related to this feature if the early breaking change is accepted in this case.

File: src/Analyzers/Core/Analyzers/Analyzers.projitems (hunk 3 above)
Comment (RikkiGibson) [reply src/Analyzers/Core/Analyzers/Analyzers.projitems:15]: No. Reverting.

File: src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:171]: I'm not sure. The safe option is definitely to just add a comment. Tagging @dotnet/roslyn-ide for guidance here.

File: src/VisualStudio/CSharp/Impl/Options/AutomationObject/AutomationObject.Style.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/VisualStudio/CSharp/Impl/Options/AutomationObject/AutomationObject.Style.cs:251]: I'll try this when I get the chance. Thanks.

File: src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs (hunk 2 above)
Comment (davidwengier) [reply src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:171]: Since this got out into a wide release, I think a comment makes sense, lest anyone added `diagnostic.IDE0190.severity = none` to their .editorconfig file.

I heard a rumour that there might have been one or two people who didn't like the feature 😉

File: src/Workspaces/CSharpTest/Formatting/FormattingTests.cs [hunk 4]
```diff
@@ -5790,7 +5790,7 @@ static object F(string s !!)
             var expectedCode =
 @"class C
 {
-    static object F(string s!!)
+    static object F(string s !!)
```
Comment (davidwengier) [inline src/Workspaces/CSharpTest/Formatting/FormattingTests.cs:5793]: Does it make sense to keep this test at all?

**Review by davidwengier — APPROVED** [review]
IDE side looks fine to me ¯\\\_(ツ)_/¯

File: src/Workspaces/CSharpTest/Formatting/FormattingTests.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Workspaces/CSharpTest/Formatting/FormattingTests.cs]: No, let's delete it.

File: src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs]: Done.

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt [hunk 5]
```diff
@@ -22,6 +22,7 @@ static Microsoft.CodeAnalysis.CSharp.SyntaxFactory.LocalFunctionStatement(Micros
...
 static Microsoft.CodeAnalysis.CSharp.SyntaxFactory.PropertyDeclaration(Microsoft.CodeAnalysis.SyntaxList<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeListSyntax!> attributeLists, Microsoft.CodeAnalysis.SyntaxTokenList modifiers, Microsoft.CodeAnalysis.CSharp.Syntax.TypeSyntax! type, Microsoft.CodeAnalysis.CSharp.Syntax.ExplicitInterfaceSpecifierSyntax? explicitInterfaceSpecifier, Microsoft.CodeAnalysis.SyntaxToken identifier, Microsoft.CodeAnalysis.CSharp.Syntax.AccessorListSyntax! accessorList) -> Microsoft.CodeAnalysis.CSharp.Syntax.PropertyDeclarationSyntax!
+*REMOVED*static Microsoft.CodeAnalysis.CSharp.SyntaxFactory.Parameter(Microsoft.CodeAnalysis.SyntaxList<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeListSyntax!> attributeLists, Microsoft.CodeAnalysis.SyntaxTokenList modifiers, Microsoft.CodeAnalysis.CSharp.Syntax.TypeSyntax? type, Microsoft.CodeAnalysis.SyntaxToken identifier, Microsoft.CodeAnalysis.SyntaxToken exclamationExclamationToken, Microsoft.CodeAnalysis.CSharp.Syntax.EqualsValueClauseSyntax? default) -> Microsoft.CodeAnalysis.CSharp.Syntax.ParameterSyntax!
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt:25]: I think we can just remove that entry, instead of marking it as removed, since it was in the "unshipped" side of things. Removal tracking is useful for "shipped" APIs

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs [hunk 6]
```diff
@@ -143,8 +143,8 @@ public enum SyntaxKind : ushort
         PercentEqualsToken = 8283,
         /// <summary>Represents <c>??=</c> token.</summary>
         QuestionQuestionEqualsToken = 8284,
-        /// <summary>Represents <c>!!</c> token.</summary>
-        ExclamationExclamationToken = 8285,
+        // Don't use 8285. It corresponds to the deleted kind ExclamationExclamationToken which was previously shipped.
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxKind.cs:146]: >shipped [](http://example.com/codeflow?start=111&length=7)

nit: It was not exactly "shipped" since it remained in preview. Consider "Don't use 8285. Was ExclamationExclamationToken before the feature was removed from C# 11"

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 7]
```diff
@@ -4520,34 +4519,23 @@ private ParameterSyntax ParseParameter()
                 //
                 // We can potentially merge that with an immediately following !! or an immediately following !=
                 // All other cases are in error.
-                var firstExclamation = this.EatToken();
+                identifier = AddTrailingSkippedSyntax(identifier, this.AddError(this.EatToken(), ErrorCode.ERR_ParameterNullCheckingNotSupported));
                 if (this.CurrentToken.Kind is SyntaxKind.ExclamationToken)
                 {
                     // have two !'s in a row.  Merge them (reporting any errors if they cannot merge properly).
-                    exclamationExclamationToken = MergeAdjacent(firstExclamation, this.EatToken(), SyntaxKind.ExclamationExclamationToken);
+                    identifier = AddTrailingSkippedSyntax(identifier, this.EatToken());
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:4526]: nit: the comment above seems out of date (we're no longer merging exclamation tokens)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 8]
```diff
@@ -11787,13 +11775,10 @@ private bool ScanExplicitlyTypedLambda(Precedence precedence)
                     }
 
                     // eat the parameter name.
-                    if (this.IsTrueIdentifier())
-                    {
-                        this.EatToken();
-                    }
+                    var identifier = this.IsTrueIdentifier() ? this.EatToken() : CreateMissingIdentifierToken();
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11778]: What's the impact of this change?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11778]: I don't expect an observable impact here. It just gives us a token to attach the `!!` to when scanning. Alternatively, we could make `ParseParameterNullCheck` accept a maybe-null identifier and just avoid calling the methods to attach the skipped tokens when the identifier is null.

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt (hunk 5 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt:25]: We did ship these APIs in 17.2.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 7 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:4523]: Updated the comments.

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 3)

**Review by jcouv — APPROVED** [review]
LGTM Thanks (new iteration 3)

**Comment by RikkiGibson** [comment]
roslyn-ci failure is a known issue addressed by #61480.
